posts/14-stable-structures.html.pm

```diff
@@ -0,0 +1,671 @@
+  Two examples of using the ◊code{StableBTreeMap::range} method.
+}
+◊source-code["rust"]{
+◊em{/// Selects all subaccounts of the specified principal.}
+fn ◊b{principal_subaccounts}(
+  balance_map: &StableBTreeMap<(Principal, Subaccount), Tokens>,
+  principal: Principal,
+) -> impl Iterator<Item = (Subaccount, Tokens)> + '_ {
+  balance_map
+    .◊b{range}((principal, Subaccount::default())..)
+    .take_while(|((p, _), )| p == principal)
+    .map(|((_, s), t)| (s, t))
+}
+
+◊em{/// Selects a transaction range for the specified principal.}
+fn ◊b{principal_tx_range}(
+  tx_index: &StableBTreeMap<(Principal, TxId), Transaction>,
+  principal: Principal,
+  start: TxId,
+  len: usize,
+) -> impl Iterator<Item = Transaction> + '_ {
+  tx_index
+    .◊b{range}((principal, start)..)
+    .take(len)
+    .map(|((_, _), tx)| tx)
+}
+}
+}
+}
+
+
+◊section{
+◊section-title["constructing-ss"]{Constructing stable structures}
+◊p{
+  We have to declare stable structures before we can use them.
+  Each data structure ◊code{T} in the library declares at least three constructors:
+}
+◊ul[#:class "arrows"]{
+  ◊li{◊code{T::new} allocates a new copy of ◊code{T} in the given ◊a[#:href "#memory"]{memory}, potentially overwriting previous memory content.}
+  ◊li{◊code{T::load} recovers a previously constructed ◊code{T} from the given ◊a[#:href "#memory"]{memory}.}
+  ◊li{◊code{T::init} is a ◊a[#:href "https://en.wikipedia.org/wiki/DWIM"]{DWIM} constructor acting as ◊code{T::new} if the given ◊a[#:href "#memory"]{memory} is empty and as ◊code{T::load} otherwise.}
+}
+◊p{
+  In practice, most canisters need only ◊code{T::init}.
+}
+}
+
+
+◊section{
+◊section-title["tying-together"]{Tying it all together}
+◊p{
+  This section contains an example of declaring several data structures for a ledger canister compatible with the ◊a[#:href "https://github.com/dfinity/ICRC-1/blob/3c64844bc6219e1d07ce05e27ec636df1e562114/standards/ICRC-1/README.md"]{ICRC-1} specification.
+  We will need three collections:
+}
+◊ol-circled{
+  ◊li{
+    A ◊code-ref["#stable-cell"]{Cell} to store the ledger metadata.
+  }
+  ◊li{
+    A ◊code-ref["#stable-btree"]{StableBTreeMap} to map accounts ◊sidenote["sn-account"]{An ◊a[#:href "https://github.com/dfinity/ICRC-1/blob/3c64844bc6219e1d07ce05e27ec636df1e562114/standards/ICRC-1/README.md#account"]{account} is a pair of a principal and a 32-byte subaccount.} to their current balance.
+  }
+  ◊li{A ◊code-ref["#stable-log"]{StableLog} for transaction history.}
+}
+◊p{
+  The ledger metadata is relatively small; it should fit into two megabytes or sixteen WebAssembly memory pages.
+  We will allocate the metadata cell in a ◊a[#:href "#restricted-memory"]{restricted memory}, partitioning the rest of the stable memory between the accounts map and the log using the ◊a[#:href "#memory-manager"]{memory manager}.
+}
+◊p{
+  Let us put these ideas to code.
+  First, we must import a few type definitions from the library.
```

Comment:
Seems like a trivial part, shall we just drop it?

Reply:
One of my goals was to ensure that if you copy all the code into a file, it compiles cleanly.

Reply:
Maybe a better approach would be to provide a link to a repo, leaving all the trivial details out. But it's fine, up to you...